### DISCUSSION TITLE
[question] New window cannot listen for emit

### DISCUSSION THREAD

**Discussion by Qxiaoqi** [tauri-apps/tauri#9303] [Q&A]
### Describe the bug

I opened a new window and then this window listens for this event. However, it cannot listen to this event the first time the window is opened. Only when the window remains open can the event be listened to.

I suspect that when the new window is first opened, it hasn't finished initializing yet, so the listener isn't active at that moment. However, the event has already been emitted.

How can I solve this problem? It has been bothering me for many days. Thank you very much.

```tsx
// TestWindowLabel
React.useEffect(() => {
  let unlisten: (() => void) | undefined = undefined
  ;(async () => {
    unlisten = await listen("change-text", async (event: Event<string>) => {
       // xxx     
    })
  })()
  return () => {
    unlisten?.()
  }
}, [])
```

```rust
// Click()
crate::windows::show_window(xxx);
handle.emit_all("change-text", "content").unwrap_or_default(),  // AppHandle
```

```rust
pub fn show_window(xxx) -> tauri::Window {
    let window = get_window(xxx);
    window.show().unwrap();
    window
}

pub fn get_window(xxx) -> tauri::Window {
    let handle = APP_HANDLE.get().unwrap();
    let window = match handle.get_window("TestWindowLabel") {
        Some(window) => {
          // xxx
        }
        None => {
            let builder = tauri::WindowBuilder::new(
                handle,
                "TestWindowLabel",
                tauri::WindowUrl::App("index.html".into()),
            )
            .title("xxx")
            .fullscreen(false)
            // ...
    
            build_window(builder)
        }
    };

    window
}


// build_window
pub fn build_window<'a, R: tauri::Runtime>(
    builder: tauri::WindowBuilder<'a, R>,
) -> tauri::Window<R> {
    {
        let window = builder
            // xxx
            .build()
            .unwrap();

        window
    }
}
```

### Full `tauri info` output

```text
[✔] Environment
    - OS: Mac OS 13.6.3 X64
    ✔ Xcode Command Line Tools: installed
    ✔ rustc: 1.74.0 (79e9716c9 2023-11-13) (Homebrew)
    ✔ cargo: 1.74.0
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: stable-x86_64-apple-darwin (default)
    - node: 18.12.1
    - pnpm: 8.5.0
    - yarn: 1.22.19
    - npm: 8.19.2
    - bun: 1.0.30

[-] Packages
    - tauri [RUST]: 1.6.1
    - tauri-build [RUST]: 1.5.1
    - wry [RUST]: 0.24.7
    - tao [RUST]: 0.16.7
    - @tauri-apps/api [NPM]: 1.5.3
    - @tauri-apps/cli [NPM]: 1.5.10 (outdated, latest: 1.5.11)

[-] App
    - build-type: bundle
    - CSP: unset
    - distDir: ../dist
    - devPath: http://localhost:1421/
    - framework: React
    - bundler: Vite
```

**Comment by venkr** [comment]
Yeah, this is an issue we've definitely run into as well.

Couple options:

1 - (Easier): Create an artificial delay between `crate::windows::show_window(xxx)` and `handle.emit_all("change-text", "content").unwrap_or_default()`. Probably something like using `tokio::time::sleep` with `tauri::async_runtime::spawn`

2 - (More reliable): If you don't see a Window, then store the message on the Rust side instead of emitting it, and then make the JS send a custom command to check for any missed initialization-time messages after its ready.

Eg: In the case of `match handle.get_window("TestWindowLabel") = None` - instead of emitting the `change-text` message, store it in `cold_start_messages`. Then register a command with `#[tauri::command]` called `get_cold_start_messages`. Finally, upon initialization inside the JS, call `get_cold_start_messages`, and process the messages received during cold start.
